hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -2028,20 +2044,33 @@ private void addBlobListAsFileStatus(final BlobList blobList,
       Path entryPath = entry.getPath();
       entryPath = entryPath.makeQualified(this.uri, entryPath);
 
-      fileStatuses.add(
-          new VersionedFileStatus(
-              owner,
-              group,
-              fsPermission,
-              hasAcl,
-              entry.getContentLength(),
-              entry.getIsDirectory(),
-              1,
-              blockSize,
-              entry.getLastModifiedTime(),
-              entryPath,
-              entry.getETag()));
+      FileStatus fileStatus = new VersionedFileStatus(
+          owner,
+          group,
+          fsPermission,
+          hasAcl,
+          entry.getContentLength(),
+          entry.getIsDirectory(),
+          1,
+          blockSize,
+          entry.getLastModifiedTime(),
+          entryPath,
+          entry.getETag());
+
+      if (entry.getETag() != null) {
+        // This is a blob entry. It is either a file or a marker blob.
+        // In both cases we will add this.
+        fileMetadata.put(blobKey, fileStatus);
+      }
+      else {
+        // This is a BlobPrefix entry. It is a directory with file inside
+        // This might have already been added as a marker blob.
+        if (!fileMetadata.containsKey(blobKey)) {
+          fileMetadata.put(blobKey, fileStatus);
+        }
```

Comment:
This is based on the assumption that the List Blob returns the entries in lexicographical order and for an explicit directory with marker blob. Blob entry will just have name whereas BlobPrefix entry will have name appended with "/".
Therefore, BlobPrefix entry will always come just after the Blob entry for an explicit directory.

Two entries with same key will only come in case of explicit directory.

Reply:
Also, to confirm this I have added tests to verify that the entry we return in case of explicit directories is of marker blob only. 
Tests for checking that the listing order is lexicographic is also added.

Reply:
Also, to confirm this I have added tests to verify that the entry we return in case of explicit directories is of marker blob only. 
Tests for checking that the listing order is lexicographic is also added.

Reply:
ok. i understand your point here!
One concern, what if the values are paginated.
Should we maintain `TreeMap<String, FileStatus> fileMetadata = new TreeMap<>();` at `listStatus` level`?

Reply:
Yes, makes sense.
taken